src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildUpstreamNodeAction.java

```diff
@@ -0,0 +1,53 @@
+package org.jenkinsci.plugins.workflow.support.steps.build;
+
+import hudson.model.InvisibleAction;
+import hudson.model.Run;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import java.util.Objects;
+
+/**
+ * Attached to newly-created builds in order to point back to the triggering FlowNode.
+ *
+ * We annotate the downstream build instead of upstream in order to support no-wait builds,
+ * because the downstream run id is not available when they're still queued.
```

Comment:
The `Queue.Item.id` _is_ available FWIW, so you _could_ place the marker on the upstream build if that were the more convenient direction to retrieve later. (You cannot jump directly from a `Queue.Item.id` to a `Run.number`, if the queue item was in fact successfully scheduled and started building; you can only check the queue ID of a given build.)

Reply:
This would be more code and much more work for the server so I think we should leave it as-is, but thanks for pointing it out, I didn't know the queue had its own id space for jobs.